library/core/src/intrinsics.rs

```diff
@@ -991,14 +995,16 @@ extern "rust-intrinsic" {
     ///
     /// let v_clone = v_orig.clone();
     ///
-    /// // The no-copy, unsafe way, still using transmute, but not relying on the data layout.
-    /// // Like the first approach, this reuses the `Vec` internals.
-    /// // Therefore, the new inner type must have the
-    /// // exact same size, *and the same alignment*, as the old type.
-    /// // The same caveats exist for this method as transmute, for
-    /// // the original inner type (`&i32`) to the converted inner type
-    /// // (`Option<&i32>`), so read the nomicon pages linked above and also
-    /// // consult the [`from_raw_parts`] documentation.
+    /// // To transmute the inner type of the contents of a container, you must make sure
+    /// // to not violate any of the container's invariants. For `Vec`, this means that
+    /// // both the size *and alignment* of the inner types match. Other containers could
+    /// // rely on different properties, even the `TypeId`, where changing the type would
+    /// // not be possible without violating the container invariants.
+    /// //
+    /// // Even though the transmute is performed through a pointer cast, the same caveats exist
+    /// // for this method as transmute, for the original inner type (`&i32`) to the converted
+    /// // inner type (`Option<&i32>`), so read the nomicon pages linked above and also consult
+    /// // the [`from_raw_parts`] documentation.
```

Comment:
I cannot entirely make sense of this paragraph (to be fair, the old text was already strange IMO). What about something like this?

```
This is the proper no-copy, unsafe way of "transmuting" a `Vec`, not relying on the data layout.
Instead of literally calling `transmute`, we perform a pointer cast, but in terms of converting
the original inner type (`&i32`) to the new one (`Option<&i32>`), this has all the same caveats.
Besides the information provided above, also consult the [`from_raw_parts`] documentation.
```

Reply:
Yeah I struggled a bit to wrap my head around the old text as well. I've updated it with your suggestion, though with slightly altered wording. Thanks.